vcd/resource_vcd_org_vdc.go

```diff
@@ -571,41 +572,162 @@ func resourceVcdVdcUpdate(d *schema.ResourceData, meta interface{}) error {
-				Enabled:      takeBoolPointer(storageConfiguration["enabled"].(bool)),
-				ProviderVdcStorageProfile: &types.Reference{
-					HREF: vdcStorageProfileDetails.ProviderVdcStorageProfile.HREF,
-				},
+			newStorageProfiles = append(newStorageProfiles, storageProfileCombo{
+				configuration: storageConfiguration,
+				reference:     nil,
 			})
-			if err != nil {
-				return fmt.Errorf("error updating VDC storage profile: %s", err)
+		}
+	}
+
+	// 3 find removed storage profiles: SP are in the VDC, but not in the definition
+	for _, vdcStorageProfile := range adminVdc.AdminVdc.VdcStorageProfiles.VdcStorageProfile {
+		found := false
+		for _, storageConfigurationValues := range set.List() {
+			storageConfiguration := storageConfigurationValues.(map[string]interface{})
+			if storageConfiguration["name"].(string) == vdcStorageProfile.Name {
+				found = true
 			}
 		}
+		if !found {
+			_, isDefault := defaultSp[vdcStorageProfile.Name]
+			if isDefault {
+				delete(defaultSp, vdcStorageProfile.Name)
+			}
+			removeStorageProfiles = append(removeStorageProfiles, storageProfileCombo{
+				configuration: nil,
+				reference:     vdcStorageProfile,
+			})
+		}
 	}
 
-	log.Printf("[TRACE] VDC update completed: %s", adminVdc.AdminVdc.Name)
-	return resourceVcdVdcRead(d, meta)
+	// 4. Check that there is one and only one default element
+	if len(defaultSp) == 0 {
+		return fmt.Errorf("no default storage profile left after update")
+	}
+	if len(defaultSp) > 1 {
+		defaultItems := ""
+		for d := range defaultSp {
+			defaultItems += " " + d
+		}
+		return fmt.Errorf("more than one default storage profile defined [%s]", defaultItems)
+	}
+
+	// 5. Update existing storage profiles
+	for _, spCombo := range existingStorageProfiles {
+		err := updateStorageProfileDetails(client, adminVdc, spCombo.reference, spCombo.configuration)
+		if err != nil {
+			return err
+		}
+	}
+
+	// 6. Add new storage profiles
+	for _, spCombo := range newStorageProfiles {
+		storageProfile, err := client.QueryProviderVdcStorageProfileByName(spCombo.configuration["name"].(string), adminVdc.AdminVdc.ProviderVdcReference.HREF)
+		if err != nil {
+			return err
+		}
+		err = adminVdc.AddStorageProfileWait(&types.VdcStorageProfileConfiguration{
+			Enabled: spCombo.configuration["enabled"].(bool),
+			Units:   "MB",
+			Limit:   int64(spCombo.configuration["limit"].(int)),
+			Default: spCombo.configuration["default"].(bool),
+			ProviderVdcStorageProfile: &types.Reference{
+				HREF: storageProfile.HREF,
+				Name: storageProfile.Name,
+			},
+		}, "")
+		if err != nil {
+			return fmt.Errorf("error adding new storage profile: %s", err)
+		}
+	}
+
+	// 7. Delete unwanted storage profiles
+	for _, spCombo := range removeStorageProfiles {
+		err := adminVdc.RemoveStorageProfileWait(spCombo.reference.Name)
+		if err != nil {
+			return fmt.Errorf("error removing storage profile %s: %s", spCombo.reference.Name, err)
+		}
+	}
```

Comment:
Have you checked how we handle case when storage profile is already use, for e.g. by VM and you try to update VDC by removing it from the list assigned to VDC?

Reply:
We don't. Here's a quote from my own comment in the issue:

> this could be a problem if the storage profile was used for any object that belongs to the VDC. My plan is - rather than over-engineering and figure out if removal is possible - to attempt removal, and let the API complain if the storage profile cannot be removed.

Reply:
That's should be ok, if API complains with reasonable message ;-) that happens not every time.